modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3230,13 +3230,16 @@ def to_pandas(self):
         if df.empty:
             df = pandas.DataFrame(columns=self.columns, index=self.index)
         else:
-            for axis in [0, 1]:
-                ErrorMessage.catch_bugs_and_request_email(
-                    not df.axes[axis].equals(self.axes[axis]),
-                    f"Internal and external indices on axis {axis} do not match.",
-                )
-            df.index = self.index
-            df.columns = self.columns
+            if self._index_cache is not None and self._columns_cache is not None:
+                # no need to check external and internal axes since in that case
+                # external axes will be computed from internal partitions
+                for axis in [0, 1]:
+                    ErrorMessage.catch_bugs_and_request_email(
+                        not df.axes[axis].equals(self.axes[axis]),
+                        f"Internal and external indices on axis {axis} do not match.",
+                    )
+                df.index = self.index
+                df.columns = self.columns
```

Comment:
We may want to preserve the logic at least partially:
```suggestion
            for axis, public_index in enumerate([self._index_cache, self._columns_cache]):
                # no need to check external and internal axes since in that case
                # external axes will be computed from internal partitions
                if public_index is not None:
                    ErrorMessage.catch_bugs_and_request_email(
                        not df.axes[axis].equals(public_index),
                        f"Internal and external indices on axis {axis} do not match.",
                    )
                    df.set_axis(axis=axis, labels=public_index, inplace=True, copy=False)
```

Reply:
good catch!